Kubeadm/dual stack support in 1.21 (#26675)
kubernetes/website: content/en/docs/concepts/services-networking/dual-stack.md

{{< /note >}}

### Enable dual-stack via kubeadm

Install kubeadm following the steps from the [Installing Kubeadm](/docs/setup/production-environment/tools/kubeadm/install-kubeadm/) documentation.

Make sure that nodes allow IPv6 forwarding, if not, run `sudo sysctl -w net.ipv6.conf.all.forwarding=1` on every node in the cluster.

{{< note >}}
`kubeadm upgrade` will change `IPv6DualStack` to true by default if feature gate is not set in old cluster. However, cluster-cidr and service-cidr modification are not supported.
{{< /note >}}

#### Create a dual-stack cluster

> A simple command with `podCidr` and `serviceCidr` flags to create a dual-stack cluster via `kubeadm init`, it like below:

Reviewer: ```suggestion
A simple command with `--pod-network-cidr` and `--service-cidr` flags to create a dual-stack cluster via `kubeadm init`, it like below:
```

```

Currently, `kube-controller-manager` flags `--node-cidr-mask-size-ipv4|--node-cidr-mask-size-ipv6` are setting with the default value. See [enable IPv4/IPv6 dual stack](/docs/concepts/services-networking/dual-stack#enable-ipv4ipv6-dual-stack).

There is a limitation here, `--apiserver-advertise-address` flag doesn't support dual-stack.

#### Join a node to dual-stack cluster

Before joining a node, make sure that the node has IPv6 routable network interface and allows IPv6 forwarding.

Here is an example kubeadm [configuration file](https://pkg.go.dev/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2) `kubeadm-config.yml` for joining a worker node to the cluster.

```yaml
apiVersion: kubeadm.k8s.io/v1beta2
kind: JoinConfiguration
discovery:
  bootstrapToken:
    apiServerEndpoint: 10.100.0.1:6443
    token: 0c0z4p.dnafh6vnmouus569
    caCertHashes: ["sha256:fcb3e956a6880c05fc9d09714424b827f57a6fdc8afc44497180905946527adf"]
```

Besides, here is an example kubeadm [configuration file](https://pkg.go.dev/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2) `kubeadm-config.yml` for joining another control plane node to the cluster.

> ```yaml

Reviewer: ```suggestion

```yaml
```

{{< /note >}}

### Enable dual-stack via kubeadm

Install kubeadm following the steps from the [Installing Kubeadm](/docs/setup/production-environment/tools/kubeadm/install-kubeadm/) documentation.

Make sure that nodes allow IPv6 forwarding, if not, run `sudo sysctl -w net.ipv6.conf.all.forwarding=1` on every node in the cluster.

{{< note >}}
`kubeadm upgrade` will change `IPv6DualStack` to true by default if feature gate is not set in old cluster. However, cluster-cidr and service-cidr modification are not supported.
{{< /note >}}

> #### Create a dual-stack cluster

Reviewer 2: Perhaps this could live inside https://kubernetes.io/docs/setup/

Author: Moved

A simple command with `--pod-network-cidr` and `--service-cidr` flags to create a dual-stack cluster via `kubeadm init`, it like below:

```shell
kubeadm init --feature-gates IPv6DualStack=true --pod-network-cidr=172.30.0.0/16,fefe:ffff:0::/48 --service-cidr=172.31.0.0/16,fefe:ffff:1::/108
```

To make things more clear, here is an example kubeadm [configuration file](https://pkg.go.dev/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2) `kubeadm-config.yml` for the dual-stack control plane node.

```yaml
apiVersion: kubeadm.k8s.io/v1beta2
kind: ClusterConfiguration
featureGates:
  IPv6DualStack: true
networking:
  podSubnet: 172.30.0.0/16,fefe:ffff:0::/48

>   serviceSubnet: 172.31.0.0/16,fefe:ffff:1::/108

Reviewer 3: lowercase is the rfc recommended, and service subnet max size is 112 IIRC

>   serviceSubnet: 172.31.0.0/16,fefe:ffff:1::/108

Author: podSubnet: 10.244.0.0/16,2001:db8:42:0::/56
  serviceSubnet: 10.96.0.0/16,2001:db8:42:1::/112

Updated

---
apiVersion: kubeadm.k8s.io/v1beta2
kind: InitConfiguration
localAPIEndpoint:
  advertiseAddress: "10.100.0.1"

>   bindPort: 6443

Reviewer: i don't object to this, but for kubeadm users are more used to 6443, because that is the default.

>   bindPort: 6443

Author: So I will keep it as 6443 or remove it ?

>   bindPort: 6443

Reviewer: my vote is to leave it 6443.

```

Besides, here is an example kubeadm [configuration file](https://pkg.go.dev/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2) `kubeadm-config.yml` for joining another control plane node to the cluster.

```yaml
apiVersion: kubeadm.k8s.io/v1beta2
kind: JoinConfiguration
controlPlane:
  localAPIEndpoint:
    advertiseAddress: "10.100.0.2"
    bindPort: 6443
discovery:
  bootstrapToken:
    apiServerEndpoint: 10.100.0.1:6443
    token: 0c0z4p.dnafh6vnmouus569
    caCertHashes: ["sha256:fcb3e956a6880c05fc9d09714424b827f57a6fdc8afc44497180905946527adf"]
```

`advertiseAddress` in JoinConfiguration.controlPlane specifies the IP address the API Server will advertise it's listening on. It equals to `--apiserver-advertise-address` flag of `kubeadm join`.

```shell
kubeadm join --config=kubeadm-config.yml ...
```

#### Create a single-stack cluster

In 1.21 the `IPv6DualStack` feature is Beta and the feature gate is defaulted to `true`. To disable the feature you must configure the feature gate to `false`. Note that once the feature is GA, the feature gate will be removed.

```shell
kubeadm init --feature-gates IPv6DualStack=false
```

> To make things more clear, here is an example kubeadm [configuration file](https://pkg.go.dev/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2) `kubeadm-config.yml` for the single-stack control plane node.

Reviewer: +1, emphasizing that is a good idea.

> To make things more clear, here is an example kubeadm [configuration file](https://pkg.go.dev/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2) `kubeadm-config.yml` for the single-stack control plane node.

Author: ```
Enabling the dual-stack feature doesn't mean that you need to use dual-stack addressing.  
You can deploy a single-stack cluster that has the dual-stack networking feature enabled.
```

Current description. Need I add more?

> To make things more clear, here is an example kubeadm [configuration file](https://pkg.go.dev/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2) `kubeadm-config.yml` for the single-stack control plane node.

Reviewer: yes, seems good.

{{< /note >}}

> ### Enable dual-stack via kubeadm

Reviewer: i'm not sure if this is the best place for this, actually.

maybe this page can just cross link to a new section under the https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/

sig-docs will likely provide feedback on this.

Reviewer: my vote here is to move this as a new page under:
https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/
and cross reference to this dual-stack.md page.

but we should wait for sig-docs feedback.

Author: > wait for sig-docs feedback

OK

Reviewer 2: For SIG Docs usually try _not_ to put task descriptions in the concepts section. We'd prefer to either create a new task page or add a section to an existing task page.

Sometimes https://kubernetes.io/docs/setup/ is a better home; that's usually for documents that only apply to brand new clusters.

Install kubeadm following the steps from the [Installing Kubeadm](/docs/setup/production-environment/tools/kubeadm/install-kubeadm/) documentation.

> Make sure that nodes allow ipv6 forwarding, if not, run `sudo sysctl -w net.ipv6.conf.all.forwarding=1` on every node in the cluster.

Reviewer: `IPv6`
please change this in all places in the new diff.

> * Notice that `kubeadm upgrade` will change IPv6DualStack to true by default if feature gate is not set in old cluster. However, cluster-cidr and service-cidr modification are not supported.

Reviewer: ```suggestion
{{< note >}}
`kubeadm upgrade` will change `IPv6DualStack` to true by default if feature gate is not set in old cluster. However, cluster-cidr and service-cidr modification are not supported.
{{< /note >}}
```

#### Create a dual-stack cluster

You must pass your `podCidr` and `serviceCidr` value to `kubeadm init`, like so:

```

> kubeadm init --feature-gates IPv6DualStack=true --pod-network-cidr=172.30.0.0/16,fefe:ffff:0::/48 --service-cidr=172.31.0.0/16,fefe:ffff:1::/108

Reviewer: i think we should just pass `--config` with the example config we discussed as the primary example target in this guide.
these kubeadm flags are GA / stable, but we recommend config to users since it's versioned.
we can still include examples with the flags as the second option if you'd like.

```

Currently, `kube-controller-manager` flags `--node-cidr-mask-size-ipv4|--node-cidr-mask-size-ipv6` are setting with the default value. See [enable IPv4/IPv6 dual stack](/docs/concepts/services-networking/dual-stack#enable-ipv4ipv6-dual-stack).

> There is a limitation here, `--apiserver-advertise-address` flag doesn't support dual-stack.

Reviewer: we should first mention the config fields:
InitConfiguration -> localAPIEndpoint
and 
JoinConfiguration -> controlPlane -> localAPIEndpoint

https://pkg.go.dev/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2

we can also mention the flag as they are GA, it's just that we recommend the config to users since it's versioned.

Author: Done

#### Join a node to dual-stack cluster

Before joining a node, make sure that the node has IPv6 routable network interface and allows IPv6 forwarding.

> 

Reviewer: ```suggestion
```

#### Create a single-stack cluster

> As IPv6DualStack feature gate is beta and default to true after 1.21, dual stack will be enabled by default without `--feature-gates` flag.

Reviewer: ```suggestion
In 1.21 the `IPv6DualStack` feature is Beta and the feature gate is defaulted to `true`. To disable the feature you must configure the feature gate to `false`. Note that once the feature is GA, the feature gate will be removed.
```

```

> kubeadm init --feature-gates IPv6DualStack=false

Reviewer: we can have the config snippet again here, but the flag version only is fine too.

Author: Done

{{< /note >}}

### Enable dual-stack via kubeadm

Install kubeadm following the steps from the [Installing Kubeadm](/docs/setup/production-environment/tools/kubeadm/install-kubeadm/) documentation.

Make sure that nodes allow IPv6 forwarding, if not, run `sudo sysctl -w net.ipv6.conf.all.forwarding=1` on every node in the cluster.

* Notice that `kubeadm upgrade` will change IPv6DualStack to true by default if feature gate is not set in old cluster. However, cluster-cidr and service-cidr modification are not supported.

#### Create a dual-stack cluster

A simple command with `podCidr` and `serviceCidr` flags to create a dual-stack cluster via `kubeadm init`, it like below:

```shell
kubeadm init --feature-gates IPv6DualStack=true --pod-network-cidr=172.30.0.0/16,fefe:ffff:0::/48 --service-cidr=172.31.0.0/16,fefe:ffff:1::/108
```

> To make things more clear, here is an example config `kubeadm-config.yml` for the dual-stack control-plane node.

Reviewer: ```suggestion
To make things more clear, here is an example kubeadm [configuration file](https://pkg.go.dev/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2) `kubeadm-config.yml` for the dual-stack control plane node.
```

someone told us to not have the dash for `control plane` in docs.
[1]

```yaml
apiVersion: kubeadm.k8s.io/v1beta2
kind: ClusterConfiguration
featureGates:
  IPv6DualStack: true
networking:

>       podSubnet: 172.30.0.0/16,fefe:ffff:0::/48
>       serviceSubnet: 172.31.0.0/16,fefe:ffff:1::/108

Reviewer: ```suggestion
  podSubnet: 172.30.0.0/16,fefe:ffff:0::/48
  serviceSubnet: 172.31.0.0/16,fefe:ffff:1::/108
```

``

> Run kubeadm to initiate the dual-stack control-plane node.

Reviewer: [1]

* [Validate IPv4/IPv6 dual-stack](/docs/tasks/network/validate-dual-stack) networking

> * [Dual-stack support with kubeadm](/docs/setup/production-environment/tools/kubeadm/dual-stack-support/)

Reviewer 2: ```suggestion
* [Enable dual-stack networking using kubeadm
](/docs/setup/production-environment/tools/kubeadm/dual-stack-support/)
```
(if making the suggested change to the title of that page)

However, this is just an example. Really I'd prefer to see “support” removed from the filename and generated URL.
